GPG Signature verification for Debian repos (#12)
uyuni-project/uyuni-rfc: 0000-deb-repo-sync-gpg-check.md

- Feature Name: deb_repo_sync_gpg_check
- Start Date: 12-06-2019
- RFC PR: (leave this empty)

# Summary

Verify GPG signature of Debian repositories when synchronizing.

# Motivation

- Why are we doing this?
- What use cases does it support?
- What is the expected outcome?

Uyuni can synchronize Debian repositories but currently it doesn't verify the GPG signature of the repo metadata.
In order to import the metadata in a secure way Uyuni must be able to verify the GPG signature of that metadata.

# Detailed design

## Database changes

> A new table is needed to store the association of a GPG key to a repository:

Reviewer: Not sure why we need this. We do already GPG metadata signature checking for SUSE repos and we do not need such an extra table. Why should we need it for debian?

Author: The idea would be to let user add GPG keys for third-party repos.

Under `Software -> Manage -> Repositories` UI there's the option to select SSL certificates for checking and/or authenticating the HTTPS URL or the repo. The SSL certs are defined under `Systems -> Autoinstallation -> GPG and SSL keys` (indeed not very intuitive). This allows uploading SSL certs or GPG keys but there's currently no way to link the GPG key to a repo (only SSL certs can be linked to a repo), hence this new table.

Reviewer: Oh, these are used to access RedHat repos. They do client certificiate authentication.
repo-sync access these values and expect a X509 SSL certificate.

Reviewer 2: There are definitely Debian repos that guard with X.509 certificates. Many of the ones I have to deal with internally are set up that way.

Reviewer: Do you have an example repo which is not signed with GPG but with X509 ?

Reviewer 2: Not one you can publicly access, unfortunately.

```
rhnContentSourceGpg
====================
content_source_id   number not null foreign key rhnContentSource(id)
gpg_key_id          number not null foreign key rhnCryptoKey(id)
created             timestamp
modified            timestamp

```

## Repo signature checking

> The GPG verification should be optional but a warning should be shown to the the user in case he's not providing information about the GPG key used to verify the metadata.

Reviewer: I would not say "is optional", instead I would say, "can be disabled". The default should always be "check it".
Making it insecure is an option an admin has but we do not recommend it.

Author: Ok, I made verification mandatory unless explicitly disabled.

```
http://mirror.example.com/ubuntu/dists/bionic-updates/InRelease
http://mirror.example.com/ubuntu/dists/bionic-updates/Release.gpg
```

There's also an alternative layout for Debian repos called "flat repos". In this case the repo directory containins both `Release*` and `Packages*` files. E.g.:
```
https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
```

_Note #1:_

In Uyuni the URL of a deb repository must point to the directory containing the binary packages for a specific architecture. E.g. for Ubuntu 18.04 updates repo is something like:
```
http://mirror.example.com/ubuntu/dists/bionic-updates/main/binary-amd64/
```
or in case of a flat repository to the directory containing both `Release*` and `Packages*` files:
```
https://download.opensuse.org/repositories/systemsmanagement:/Uyuni:/Master:/Ubuntu1804-Uyuni-Client-Tools/xUbuntu_18.04/
```

_Note #2:_

Debian repos don't have a standard location for the public GPG keys used to verify the metadata. Sometimes the key is located in the root of a flat repo but the location is not standardized.

### GPG keys management

> Uyuni ships the SUSE and OpenSUSE keys in the package `uyuni-build-keys`. The keys for Ubuntu and Debian should be shipped in this package too.

Reviewer: What if a user do not want to trust such a key? I would suggest to keep the automatically trusted keys very limited.
An option could be to create sub-packages and the admin can decide which keys he want to trust by install them or not.

Author: Changed to ship Debian and Ubuntu keys in separate packages.

> Any additional keys that need to be used (e.g. for Ubuntu derivatives distros or for custom repos) must be created manually using the existing UI (`Systems -> Autoinstallation -> GPG and SSL keys`)

Reviewer: spacewalk-repo-sync ask to trust the key and if the user answer with "yes" it imports it.
For Debian/Ubuntu the key may not exist at a specific place so we have a problem to download / guess it.
But maybe we could ask a keyserver?

`Systems -> Autoinstallation -> GPG and SSL keys` is a place where Keys gets stored for autoinstallation.
Sometimes they are used also for different use cases, but it is not very intuitive .

Reviewer 2: Note that [`distribution-gpg-keys`](https://github.com/xsuchy/distribution-gpg-keys) package in Fedora and openSUSE ships many of the RPM distribution GPG keys.

In Fedora, we ship the Debian keys in `debian-keyring` package. The Ubuntu keys are shipped in the `ubu-keyring` package.

We may want to standardize on this approach and unify all the keys we ship this way...

Author: Please see also my previous [comment](https://github.com/uyuni-project/uyuni-rfc/pull/12#discussion_r322756074). 

I think there could be several options for adding new GPG keys for third party repos:

1) The UI (`Systems -> Autoinstallation -> GPG and SSL keys`). Not very intuitive. Maybe we could move it to some other place in the menu.

2) Let the user fiddle directly with `/var/lib/spacewalk/gpgdir/pubring.gpg`. I think this is risk.

3) Write a small shell script to let the user import or remove third party gpg keys into `/var/lib/spacewalk/gpgdir/pubring.gpg`.

### GPG Key usage in the UI

The UI for creating/updating a repository (`Software -> Manage -> Repositories`) must be extended with one additional field:
- GPG key

> This field must be visible only if the selected repository type is `deb` and the checkbox `Has Signed Metadata` is checked. For `deb` repositories the `Has Signed Metadata` checkbox must be enabled by default.

Reviewer: This suggest now a different way compared to the way a section before.
And it does it only for debian. Why not for RPM as well?

But this is new UI and different key handling as we have it now.
Do we really need to go that deep for a first step? Or is this already step 2?

Author: As a first step we could manage with shipping Ubuntu and Debian official keys in packages. I guess that's the main use case anyways and leave custom repos for a second phase.

- `http://mirror.example.com/$REPODIR/` in the case of flat repositories

Looking up the `Release` file should be done automatically when syncing of the repo. An error should be shown if it's not possible to locate the `Release` file and its corresponding signature.

### Metadata Verification

Uyuni should verify the integrity of the `Packages.gz|xz` files using the checksums present in the `Release` file.

### Signature verification

If the `InRelease` or `Release.pgg` exist in the repo to be synced then Uyuni must verify the `Release` file. If the repo has signed metadata enabled but no GPG key selected, the default keyring must be used:
```
gpgv --keyring /var/lib/spacewalk/gpgdir/pubring.gpg InRelease
```
or
```
gpgv --keyring /var/lib/spacewalk/gpgdir/pubring.gpg Release.gpg Release
```

Otherwise the custom GPG key associated with the repo must be used:
```
gpgv --keyring </path/to/temp/keyring/file.gpg> InRelease
```
or
```
gpgv --keyring </path/to/temp/keyring/file.gpg> Release.gpg Release
```

The signature verification should be done during repo syncing when the metadata is parsed.

> Verification must be done only if the user has configured the GPG signing of the repos to be synced. Otherwise a warning must be printed in the logs about the lack of GPG configuration.

Reviewer: I would turn around this sentence as "verify" should be the default. 

```suggestion
If the user has turned off GPG signature verification, a warning should be printed into the logs.
```

# Drawbacks
[drawbacks]: #drawbacks

No drawbacks.

# Alternatives
[alternatives]: #alternatives

> - Support only a very limited set of additional fields like `Multi-Arch` while discarding all others.

Reviewer: this is an alternative for a different RFC.

Author: Removed.

# Unresolved questions
[unresolved]: #unresolved-questions

> - Should all the fields be stored ? Should only the fields defined in [1] be stored ? If yes what happens with user defined fields [2]

Reviewer: this is a question of a different RFC.

Author: removed

> - In case no GPG keys are configured for a repo (i.e. no GPG verification needed) should the `Release` file still be looked up and the checksums verified ?

Reviewer: I would suggest "yes". It would prevent using inconsistent repositories.

Reviewer 2: `Release` files are going away from Debian repositories. We need to start generating and validating `InRelease` files instead.
